File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/EventNotificationSample/src/com/hp/workpath/sample/eventnotificationsample/util/PreferenceManager.java
package com.hp.workpath.sample.eventnotificationsample.util;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONException;
import org.json.JSONObject;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class PreferenceManager {

    private static final String PREFS_NAME = "EventPrefs";

    private static final String KEY_INDEX = "keyIndex";

    public static final String DATE = "date";

    public static final String EVENT = "event";

    public static final String DATA = "data";

    static public void saveEvent(Context context, String event, String data) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        String time = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss").format(LocalDateTime.now());
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put(DATE, time);
            jsonObject.put(EVENT, event);
            if (data != null && !data.isEmpty())
                jsonObject.put(DATA, data);
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
        int keyIndex = prefs.getInt(KEY_INDEX, 0);
        editor.putString(String.valueOf(keyIndex), jsonObject.toString());
        editor.putInt(KEY_INDEX, keyIndex + 1).apply();
    }

    static public List<JSONObject> getAllEvents(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        Map<String, ?> allEntries = prefs.getAll();
        List<Integer> sortedKeys = new ArrayList<>();

        for (String key : allEntries.keySet()) {
            try {
                sortedKeys.add(Integer.parseInt(key));
            } catch (NumberFormatException e) {
                // ignore
            }
        }
        Collections.sort(sortedKeys, Collections.reverseOrder());

        List<JSONObject> events = new ArrayList<>();
        for (int key: sortedKeys) {
            try {
                events.add(new JSONObject(prefs.getString(String.valueOf(key), "")));
            } catch (JSONException e) {
                // ignore
            }
        }
        return events;
    }

    static public void registerPreferenceChangeListener(Context context, SharedPreferences.OnSharedPreferenceChangeListener listener) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        prefs.registerOnSharedPreferenceChangeListener(listener);
    }

    static public void unregisterPreferenceChangeListener(Context context, SharedPreferences.OnSharedPreferenceChangeListener listener) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        prefs.unregisterOnSharedPreferenceChangeListener(listener);
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/EventNotificationSample/src/com/hp/workpath/sample/eventnotificationsample/util/PreferenceManager.kt
package com.hp.workpath.sample.eventnotificationsample.util

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONException
import org.json.JSONObject
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

class PreferenceManager {
    companion object {
        private const val PREFS_NAME = "EventPrefs"

        private const val KEY_INDEX = "keyIndex"

        const val DATE = "date"

        const val EVENT = "event"

        const val DATA = "data"

        @JvmStatic
        fun saveEvent(context: Context?, event: String?, data: String?) {
            val prefs = context?.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val editor = prefs?.edit()
            val time = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss").format(LocalDateTime.now())
            val jsonObject = JSONObject()
            try {
                jsonObject.put(DATE, time)
                jsonObject.put(EVENT, event)
                if (!data.isNullOrEmpty()) {
                    jsonObject.put(DATA, data)
                }
            } catch (e: JSONException) {
                throw RuntimeException(e)
            }
            val keyIndex = prefs?.getInt(KEY_INDEX, 0)
            editor?.putString(keyIndex.toString(), jsonObject.toString())
            if (keyIndex != null) {
                editor?.putInt(KEY_INDEX, keyIndex + 1)?.apply()
            }
        }

        @JvmStatic
        fun getAllEvents(context: Context): List<JSONObject> {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val allEntries = prefs.all
            val sortedKeys = mutableListOf<Int>()

            for (key in allEntries.keys) {
                try {
                    sortedKeys.add(key.toInt())
                } catch (e: NumberFormatException) {
                    // ignore
                }
            }
            sortedKeys.sortDescending()

            val events = mutableListOf<JSONObject>()
            for (key in sortedKeys) {
                try {
                    events.add(JSONObject(prefs.getString(key.toString(), "") ?: ""))
                } catch (e: JSONException) {
                    // ignore
                }
            }
            return events
        }

        @JvmStatic
        fun registerPreferenceChangeListener(context: Context, listener: SharedPreferences.OnSharedPreferenceChangeListener) {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            prefs.registerOnSharedPreferenceChangeListener(listener)
        }

        @JvmStatic
        fun unregisterPreferenceChangeListener(context: Context, listener: SharedPreferences.OnSharedPreferenceChangeListener) {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            prefs.unregisterOnSharedPreferenceChangeListener(listener)
        }
    }
}